bootloaders|tests/riotboot: broken with BUILD_IN_DOCKER and wrong flashfile

#### Description

The application in `bootloaders/riotboot` has the following issues, from a first look I think I also found the culprits:

* Introduced by bb71e979f
  * the file use when flashing is not the bootloader firmware but an extended bootloader with a slot0 bootloader. So `riotboot/flash-bootloader` would overwrite `slot0` and `slot1` header.
* Introduced by aaa187e87 Fixed by https://github.com/RIOT-OS/RIOT/pull/12302
  * it does not compile with `BUILD_IN_DOCKER` anymore
  * it uses the local toolchain

The application in `tests/riotboot` also has the same `BUILD_IN_DOCKER` issues.

The `tests/riotboot` test does not work with `BUILD_IN_DOCKER`.

#### Steps to reproduce the issue

<details><summary>FLASHFILE is extended file</summary>

```
BOARD=iotlab-m3 make --no-print-directory -C bootloaders/riotboot/ info-debug-variable-FLASHFILE
/home/harter/work/git/RIOT/bootloaders/riotboot/bin/iotlab-m3/riotboot-slot0-extended.bin
```
</details>


<details><summary>tests/riotboot test does not work with `BUILD_IN_DOCKER=1`</summary>

```
WARNING:nrf52dk.tests/riotboot:make RIOT_CI_BUILD=1 CC_NOCOLOR=1 --no-print-directory -C ./tests/riotboot test
expecting slot number 0, app_ver 0
curslotnr
/srv/ilab-builds/workspace/jobs/experimental-pull-request-tests/dist/tools/pyterm/pyterm -p "/dev/riot/tty-nrf52dk" -b "115200" --noprefix --no-repeat-command-on-empty-line
Twisted not available, please install it if you want to use pyterm's JSON capabilities
Connect to serial port /dev/riot/tty-nrf52dk
Welcome to pyterm!
Type '/exit' to exit.
main(): This is RIOT! (Version: buildtest)
Hello riotboot!
You are running RIOT on a(n) nrf52dk board.
This board features a(n) nrf52 MCU.
riotboot_test: running from slot 0
Image magic_number: 0x544f4952
Image Version: 0x00000000
Image start address: 0x00002400
Header chksum: 0xa57ac1a1

>  curslotnr
Current slot=0
> curslothdr
 curslothdr
Image magic_number: 0x544f4952
Image Version: 0x00000000
Image start address: 0x00002400
Header chksum: 0xa57ac1a1

> getslotaddr 0
 getslotaddr 0
Slot 0 address=0x00002400
> dumpaddrs
 dumpaddrs
slot 0: metadata: 0x2000 image: 0x00002400
slot 1: metadata: 0x41000 image: 0x00000000
> 
compiling /srv/ilab-builds/workspace/jobs/experimental-pull-request-tests/dist/tools/riotboot_gen_hdr/bin/genhdr...
make: Nothing to be done for 'all'.
make[1]: *** No rule to make target '/srv/ilab-builds/workspace/jobs/experimental-pull-request-tests/tests/riotboot/bin/nrf52dk/tests_riotboot-slot1.bin', needed by '/srv/ilab-builds/workspace/jobs/experimental-pull-request-tests/tests/riotboot/bin/nrf52dk/tests_riotboot-slot1.hdr'.  Stop.
/srv/ilab-builds/workspace/jobs/experimental-pull-request-tests/tests/riotboot/../../Makefile.include:669: recipe for target 'test' failed
make: *** [test] Error 1
```
</details>


##### Fixed

<details><summary>It does not build with `docker`</summary>

```
BUILD_IN_DOCKER=1 DOCKER="sudo docker" BOARD=frdm-k64f make --no-print-directory -C bootloaders/riotboot
arm-none-eabi-gcc: error: /home/harter/work/git/RIOT/bootloaders/riotboot/bin/frdm-k64f/cpu/fcfield.o: No such file or directory
arm-none-eabi-gcc: error: /home/harter/work/git/RIOT/bootloaders/riotboot/bin/frdm-k64f/cpu/vectors.o: No such file or directory
arm-none-eabi-gcc: error: /home/harter/work/git/RIOT/bootloaders/riotboot/bin/frdm-k64f/newlib_syscalls_default/syscalls.o: No such file or directory
arm-none-eabi-gcc: error: /home/harter/work/git/RIOT/bootloaders/riotboot/bin/frdm-k64f/application_riotboot.a: No such file or directory
arm-none-eabi-gcc: error: /home/harter/work/git/RIOT/bootloaders/riotboot/bin/frdm-k64f/board.a: No such file or directory
arm-none-eabi-gcc: error: /home/harter/work/git/RIOT/bootloaders/riotboot/bin/frdm-k64f/checksum.a: No such file or directory
arm-none-eabi-gcc: error: /home/harter/work/git/RIOT/bootloaders/riotboot/bin/frdm-k64f/core.a: No such file or directory
arm-none-eabi-gcc: error: /home/harter/work/git/RIOT/bootloaders/riotboot/bin/frdm-k64f/cortexm_common.a: No such file or directory
arm-none-eabi-gcc: error: /home/harter/work/git/RIOT/bootloaders/riotboot/bin/frdm-k64f/cortexm_common_periph.a: No such file or directory
arm-none-eabi-gcc: error: /home/harter/work/git/RIOT/bootloaders/riotboot/bin/frdm-k64f/cpu.a: No such file or directory
arm-none-eabi-gcc: error: /home/harter/work/git/RIOT/bootloaders/riotboot/bin/frdm-k64f/newlib_syscalls_default.a: No such file or directory
arm-none-eabi-gcc: error: /home/harter/work/git/RIOT/bootloaders/riotboot/bin/frdm-k64f/periph.a: No such file or directory
arm-none-eabi-gcc: error: /home/harter/work/git/RIOT/bootloaders/riotboot/bin/frdm-k64f/riotboot.a: No such file or directory
arm-none-eabi-gcc: error: /home/harter/work/git/RIOT/bootloaders/riotboot/bin/frdm-k64f/stdio_uart.a: No such file or directory
arm-none-eabi-gcc: error: /home/harter/work/git/RIOT/bootloaders/riotboot/bin/frdm-k64f/sys.a: No such file or directory
arm-none-eabi-gcc: error: /home/harter/work/git/RIOT/bootloaders/riotboot/bin/frdm-k64f/periph_common.a: No such file or directory
/home/harter/work/git/RIOT/makefiles/boot/riotboot.mk:35: recipe for target '/home/harter/work/git/RIOT/bootloaders/riotboot/bin/frdm-k64f/riotboot-slot0.elf' failed
```
</details>

<details><summary>Building uses the local toolchain</summary>

```
PATH=${DEFAULT_PATH} BUILD_IN_DOCKER=1 DOCKER="sudo docker" BOARD=frdm-k64f make --no-print-directory -C bootloaders/riotboot/  
/bin/sh: 1: arm-none-eabi-gcc: not found
/home/harter/work/git/RIOT/makefiles/boot/riotboot.mk:35: recipe for target '/home/harter/work/git/RIOT/bootloaders/riotboot/bin/frdm-k64f/riotboot-slot0.elf' failed
make: *** [/home/harter/work/git/RIOT/bootloaders/riotboot/bin/frdm-k64f/riotboot-slot0.elf] Error 127
```
</details>

#### Expected results

* `FLASHFILE` is `${BINDIR}/riotboot.elf`
* It compiles correctly in docker without local toolchain

#### Versions

Found on e5fe8684eae4f8e139800b94b4689f0e238eb816

Commit introducing the issues bb71e979f and aaa187e87

--- a/Makefile.include
+++ b/Makefile.include
@@ -442,6 +442,9 @@ ifeq (,$(FLASHFILE))
   $(error FLASHFILE is not defined for this board: $(FLASHFILE))
 endif
 
+# By default always build ELFFILE and FLASHFILE
+BUILD_FILES += $(ELFFILE) $(FLASHFILE)
+
 # variables used to compile and link c++
 CPPMIX ?= $(if $(wildcard *.cpp),1,)
 
@@ -470,7 +473,7 @@ ifeq ($(BUILD_IN_DOCKER),1)
 link: ..in-docker-container
 else
 ifeq (,$(RIOTNOLINK))
-link: ..compiler-check ..build-message $(ELFFILE) $(FLASHFILE) print-size
+link: ..compiler-check ..build-message $(BUILD_FILES) print-size
 else
 link: ..compiler-check ..build-message $(BASELIBS)
 endif # RIOTNOLINK

--- a/makefiles/boot/riotboot.mk
+++ b/makefiles/boot/riotboot.mk
@@ -49,9 +49,7 @@ SLOT_RIOT_ELFS = $(BINDIR_APP)-slot0.elf $(BINDIR_APP)-slot1.elf
 # ensure both slot elf files are always linked
 # this ensures that both "make test" and "make test-murdock" can rely on them
 # being present without having to trigger re-compilation.
-ifneq (1, $(RIOTNOLINK))
-link: $(SLOT_RIOT_ELFS)
-endif
+BUILD_FILES += $(SLOT_RIOT_ELFS)
 
 # Create binary target with RIOT header
 $(SLOT_RIOT_BINS): %.$(APP_VER).riot.bin: %.hdr %.bin

--- a/makefiles/vars.inc.mk
+++ b/makefiles/vars.inc.mk
@@ -80,6 +80,7 @@ export ELFFILE               # The unstripped result of the compilation.
 export HEXFILE               # The 'intel hex' stripped result of the compilation.
 # BINFILE                    # The 'binary' stripped result of the compilation.
 # FLASHFILE                  # The output file used for flashing
+# BUILD_FILES                # The list of files to be built
 # DEBUGGER                   # The command to call on "make debug", usually a script starting the GDB front-end.
 # DEBUGGER_FLAGS             # The parameters to supply to DEBUGGER.
 # DEBUGSERVER                # The command to call on "make debug-server", usually a script starting the GDB server.

